src/relay/pass/quantize_rewrite.cc

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file quantize_rewrite.cc
+ * \brief Lower quantized ops to exisiting Relay ops.
+ */
+
+#include <tvm/relay/pass.h>
+#include <tvm/relay/op_attr_types.h>
+#include <tvm/relay/attrs/nn_quantize.h>
+#include "pattern_util.h"
+
+namespace tvm {
+namespace relay {
+
+Expr QuantizedConv2DForwardRewrite(const Call& ref_call,
+                                   const Array<Expr>& new_args,
+                                   const NodeRef& ctx) {
+  // TODO(janimesh) - This is not the right calculation. This only serves as a
+  // prototype to discuss the flow of lowering of quantization ops and
+  // namespaces.
+  CHECK_EQ(new_args.size(), 2);
+  Expr quantized_data = new_args[0];
+  Expr quantized_kernel = new_args[1];
+  const auto* param = ref_call->attrs.as<QuantizedConv2DAttrs>();
+  CHECK_EQ(param->input_zero_point, 0) << "Only symmetric support yet";
+  CHECK_EQ(param->kernel_zero_point, 0) << "Only symmetric support yet";
+  CHECK_EQ(param->output_zero_point, 0) << "Only symmetric support yet";
+  // TODO(janimesh) - The out_dtype should be something else, like "int32".
+  Expr int8_conv = Conv2D(quantized_data,
```

Comment:
Where is our `data - input_zero_point` / `kernel - kernel_zero_point` computation? into `quantized_data` / `quantized_kernel`? Because we will cast data / kernel be int16/ int32 to minus zero point, I care about this.

Reply:
There are a couple of points that can answer your question here

* Currently, the rewrite of quantized_conv2d to a sequence of Relay operation is not complete. I was hoping to start the discussion of quantization flow, namespaces before we can start discussing the details of operations. I will work on this asynchronously, as I understand the computation. Please let me know more of your thoughts on the quantization flow, like namespaces, the infra usage for lowering the quantized ops into a series of Relay ops.

* First PR would only support symmetric quantization. The asymmetric computation that has the `data - input_zero_point` kernel will come later, one people agree on the quantization flow.  See Line 44 - 46 that are limiting the scope to symmetric quantization. Highly possible, that I will need you help in lowering the asymmetric quantization convolution. But, we should not complicate this PR by including asymmetric quantization.

Reply:
Ok. Haven't noticed line 44 - 46 comment before.